### ISSUE TITLE
onServer() helper

### ISSUE THREAD

**Issue by lcswillems** [zeit/next.js#8804]
# Feature request

## Is your feature request related to a problem? Please describe.

It would be nice to have a built-in helper to detect if code is run on server or on client.

## Describe the solution you'd like

```
function onServer() {
    return typeof window === 'undefined'
}
```

**Comment by timneutkens** [comment]
The best thing you can do is adding this exact check `typeof window === 'undefined'` as it optimizes and tree shakes away the code.

**Comment by lcswillems** [comment]
Sorry, but I don't understand your answer. I already have this snippet in my codebase. I ask to add it directly to Next.js one. Because everybody wants this and I saw a lot of people adding it to their codebase too.

I don't understand what you suggest me and why you close this.

**Comment by timneutkens** [comment]
```
if(typeof window === 'undefined') {
  // some code
}
```

Turns into this on the client-side code:

```
if(false) {
  // some code
}
```

And is tree shaken away so ends up being removed from the client-side bundle.

Increasing the api surface would make these optimizations significantly harder and non-standard.

**Comment by lcswillems** [comment]
Okay, thank you for the explanations!

I don't know the tools you are using so I don't understand why it would be harder to optimize. On client, `isServer()` will also always return false...

**Comment by timneutkens** [comment]
It's not about returning, it's about removing the exact code, you don't want to send huge libraries that you import inside the if statement to be sent to the browser for example.

**Comment by lcswillems** [comment]
Okay. But if you are able to detect that `typeof window === 'undefined'` is always false, you will detect that it is:

```
if (false) {
  // some code
}
```

and be able to remove this code, and you won't send the libraries like in the previous case. From my point of view, I don't see the difference. But it is maybe because I have badly understand something.

Moreover, you are always sending `next.js` to client right? So adding this little function won't change anything, but I might be wrong.

**Comment by timneutkens** [comment]
It would increase:
- bundle sizes
- js parse time

Whereas `typeof window === 'undefined'` allows you to write universally working code that doesn't require anything Next.js specific.

**Comment by lcswillems** [comment]
> It would increase:
> 
> * bundle sizes
> * js parse time

I don't think it is a relevant argument. If you don't want to increase bundle size and js parse time, stop adding things to Next.js, or even better, remove everything from Next.js, you will have the best possible bundle size and parse time.

> Whereas `typeof window === 'undefined'` allows you to write universally working code that doesn't require anything Next.js specific.

I don't think it is a relevant argument too. If I copy paste all the code of Next.js in my project, I will have fully universally working code that doesn't require anything Next.js specific... But this is non-sense.

The purpose of a framework is to centralize and organize code that everybody needs for a specific task. This is part of the code that everybody needs, or almost everybody. I use it 10 times in a small project, I saw it defined in several Next.js packages. It makes sense for me to include it in Next.js.

Moreover, it makes:
- development faster because less code to write and no need to Google for a long time to find this snippet
- source code easier to understand because it is not obvisous that `typeof window === 'undefined'` detects whether client or server

I don't understand why it is not included in Next.js, as I don't understand why you closed this issue, and why you closed it so fast.

**Comment by timneutkens** [comment]
I've explained already that these 2 methods work vastly different and that your assumptions about bundle size are wrong.

> I don't think it is a relevant argument. If you don't want to increase bundle size and js parse time, stop adding things to Next.js, or even better, remove everything from Next.js, you will have the best possible bundle size and parse time.

We are, we only ship code that is actively being used and try to compile away the rest. The `typeof window` is part of that.

Note that in general you shouldn't be using checks like this and use the React lifecycle itself to handle calling methods at the appropriate time. For example `useEffect` is only called on the client-side.

`typeof window` is an optimization mostly used for library code and to tree shake away code blocks.

> as I don't understand why you closed this issue

Because we're not going to add in this feature for the reason I mentioned before.

**Comment by lcswillems** [comment]
Okay, thank you for your answer.

> we only ship code that is actively being used and try to compile away the rest. I think it is the case of the snippet. Or a lot of people use similar alternatives.

You suggest me to use `useEffect`. But, if I have this kind of code:

```
if (typeof window === 'undefined') {
    // do this
} else {
   // do that
}
```

How do you do with `useEffect`?

**Comment by timneutkens** [comment]
```
function MyComponent() {
  const [mounted, setMounted] = useState(false)

  useEffect(() => setMounted(true), [])
  return <h1>Has been mounted? {mounted ? 'yes' : 'no'}</h1>
}

```

**Comment by lcswillems** [comment]
Thank you for the example.

But this only works when the condition is in a render function. And it is much more complicated than the code before. It requires using states, etc...

Anyway thank you for the example.

I think it would be a good idea to add documentation about how to execute a specific code on server and another one on client, along with `typeof window === 'undefined'` because I think it is a frequent issue newbies have.

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
